scalableminds/webknossos
# Add banner for scheduled maintenance

### URL of deployed dev instance (used for testing):
- https://___.webknossos.xyz

### Steps to test:
Backend:
- Create new maintenance with `curl -X POST 'http://localhost:9000/api/maintenances' -H 'X-Auth-Token: secretSampleUserToken' -H "Content-Type: application/json" -d '{"startTime":1692272319000, "endTime":1692272320000, "message":"a message!"}'` (adapt timestamps to something in the future or current, use https://www.epochconverter.com/ for millisecond unix timestamps)
- delete and update routes, as well as readOne and listAll are also only available for super users
  - `curl -X DELETE 'http://localhost:9000/api/maintenances/64de02abee0100c103ab66b6' -H 'X-Auth-Token: secretSampleUserToken' -H` (replace id)
  - `curl -X PUT 'http://localhost:9000/api/maintenances/64de0390ee0100a006ab66cc' -H 'X-Auth-Token: secretSampleUserToken' -H "Content-Type: application/json" -d '{"startTime":1692272319000,  "message":"a message!!!"}'`
- You should see maintenances at `GET api/maintenances/listCurrentAndUpcoming` (this is what the frontend uses) 
Frontend:
- open wk without any maintenances -> should look as always
- create an upcoming maintenance -> wk should show this in a banner. the banner should be closable. a page refresh should not show that banner anymore.
- create a current maintennance -> wk should show this banner, but it should not be closable.
- scrolling with the banner should behave intuitively.
- in the annotation view, touching the banner with the mouse should move the banner to the bottom of the page (and vice versa)

### TODOs:
  - [x] schema
  - [x] clarify permissions
  - [x] CRUD routes
  - [x] Adapt to new format of maintenance rest api route (`GET api/maintenances/listCurrentAndUpcoming`)
  - [x] Banner for closest upcoming maintenance “Maintenance starting in 3h” with tooltip showing more details
  - [x] Banner for current maintenance “Current Maintenance” with tooltip showing more details

### Issues:
- fixes #7257

------
(Please delete unneeded items, merge only when none are left open)
